alphadia/fdrx/stats.py

```diff
@@ -0,0 +1,166 @@
+
+    Returns
+    -------
+
+    np.ndarray
+        The PEP values with same shape and order as the input dataframe.
+
+    """
+
+    score_bins = np.linspace(0, 1, pep_granularity)
+    target_decoy = psm_df[decoy_column].values
+    score = psm_df[score_column].values
+
+    target_hist, _ = np.histogram(score[target_decoy == 0], bins=score_bins)
+    decoy_hist, _ = np.histogram(score[target_decoy == 1], bins=score_bins)
+
+    std_norm = score_std / (score_bins[1] - score_bins[0])
+    # create a gaussian kernel of 0.01 width with 5 elements
+    kernel_gaussian = np.exp(
+        -(np.arange(-kernel_size, kernel_size + 1) ** 2) / (2 * std_norm**2)
+    )
+
+    # convolve the target and decoy histograms with the kernel
+    target_hist = np.convolve(target_hist, kernel_gaussian, mode="same")
+    decoy_hist = np.convolve(decoy_hist, kernel_gaussian, mode="same")
+
+    # numerical stability
+    EPSILON = 1e-6
+    pep = decoy_hist / (target_hist + decoy_hist + EPSILON)
+
+    return pep[np.digitize(score, score_bins) - 1]
+
+
+def _get_q_values(
+    _df: pd.DataFrame,
+    score_column: str = "decoy_proba",
+    decoy_column: str = "decoy",
+    qval_column: str = "qval",
+    r_target_decoy: float = 1.0,
+):
+    """Calculates q-values for a dataframe containing PSMs.
+
+    Parameters
+    ----------
+
+    _df : pd.DataFrame
+        The dataframe containing the PSMs.
+
+    score_column : str, default='proba'
+        The name of the column containing the score to use for the selection.
+        Ascending sorted values are expected.
+
+    decoy_column : str, default='_decoy'
+        The name of the column containing the decoy information.
+        Decoys are expected to be 1 and targets 0.
+
+    qval_column : str, default='qval'
+        The name of the column to store the q-values in.
+
+    Returns
+    -------
+
+    pd.DataFrame
+        The dataframe containing the q-values in column qval.
+
+    """
+    EPSILON = 1e-6
+    _df = _df.sort_values([score_column, score_column], ascending=True)
+    target_values = 1 - _df[decoy_column].values
+    decoy_cumsum = np.cumsum(_df[decoy_column].values)
+    target_cumsum = np.cumsum(target_values)
+    fdr_values = decoy_cumsum / (target_cumsum + EPSILON)
+    _df[qval_column] = _fdr_to_q_values(fdr_values) * r_target_decoy
+    return _df
+
+
+def _fdr_to_q_values(fdr_values: np.ndarray):
+    """Converts FDR values to q-values.
+    Takes a ascending sorted array of FDR values and converts them to q-values.
+    for every element the lowest FDR where it would be accepted is used as q-value.
+
+    Parameters
+    ----------
+    fdr_values : np.ndarray
+        The FDR values to convert.
+
+    Returns
+    -------
+    np.ndarray
+        The q-values.
+    """
+    fdr_values_flipped = np.flip(fdr_values)
+    q_values_flipped = np.minimum.accumulate(fdr_values_flipped)
+    q_vals = np.flip(q_values_flipped)
+    return q_vals
+
+
+def _keep_best(
+    df: pd.DataFrame,
```

Comment:
more speaking name? you can then use `df` instead of `temp_df` within the method